src/lib.rs

```diff
@@ -256,6 +257,99 @@ impl FailPoint {
     }
 }
 
+#[derive(Default)]
+struct FailPointRegistry {
+    // TODO: remove rwlock or store *mut FailPoint
+    registry: RwLock<HashMap<String, Arc<FailPoint>>>,
+}
+
+lazy_static! {
+    static ref REGISTRY: FailPointRegistry = FailPointRegistry::default();
+}
+
+pub fn setup() {
+    let mut registry = REGISTRY.registry.write().unwrap();
+    let failpoints = env::var("FAILPOINTS").unwrap();
+    for mut cfg in failpoints.trim().split(";") {
+        cfg = cfg.trim();
+        if cfg.trim().is_empty() {
+            continue;
+        }
+        let (name, order) = partition(cfg, '=');
+        match order {
+            None => panic!("invalid failpoint: {:?}", cfg),
+            Some(order) => set(&mut registry, name.to_owned(), order).unwrap(),
+        }
+    }
+}
+
+pub fn teardown() {
+    let mut registry = REGISTRY.registry.write().unwrap();
+    for (_, p) in &*registry {
+        // wake up all pause failpoint.
+        p.set_actions(vec![]);
+    }
+    registry.clear();
+}
+
+pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
+    let p = {
+        let registry = REGISTRY.registry.read().unwrap();
+        match registry.get(name) {
+            None => return None,
+            Some(p) => p.clone(),
+        }
+    };
+    p.eval(name).map(f)
+}
+
+pub fn cfg<S: Into<String>>(name: S, orders: &str) -> Result<(), String> {
+    let mut registry = REGISTRY.registry.write().unwrap();
+    set(&mut registry, name.into(), orders)
+}
+
+pub fn remove<S: AsRef<str>>(name: S) {
+    let mut registry = REGISTRY.registry.write().unwrap();
+    if let Some(p) = registry.remove(name.as_ref()) {
+        // wake up all pause failpoint.
+        p.set_actions(vec![]);
+    }
+}
+
+fn set(
+    registry: &mut HashMap<String, Arc<FailPoint>>,
+    name: String,
+    orders: &str,
+) -> Result<(), String> {
+    // `orders` are in the format of `failpoint[->failpoint...]`.
+    let actions = try!(orders.split("->").map(Action::from_str).collect());
+    // Please note that we can't figure out whether there is a failpoint named `name`,
+    // so we may insert a failpoint that doesn't exist at all.
+    let p = registry
+        .entry(name)
+        .or_insert_with(|| Arc::new(FailPoint::new()));
+    p.set_actions(actions);
+    Ok(())
+}
+
+#[macro_export]
+macro_rules! fail_point {
+    ($name:expr, $e:expr) => {{
+        let name = concat!(module_path!(), "::", $name);
+        if let Some(res) = $crate::eval(name, $e) {
+            return res;
+        }
+    }};
+    ($name:expr) => {{
+        fail_point!($name, |_| {});
```

Comment:
Does it only fit in functions that return nothing? `Fn<Args>`

Reply:
Yes.